DEVELOPER_WORKFLOW.md

```diff
@@ -0,0 +1,191 @@
+
+<img width='936' height='827' src='branding/Branches.png'/>
+
+## Recommended branch usage
+* use unity/master+sdk/master (or sdk/green) for quick tests/debugging released MREs.
+* use unity/master(or altspacevr retail build)+sdk/green for minor node fixes/hotfixes
+* use unity/red+sdk/red for general development.
+
+## Misc. Notes
+
+(Once we guarantee back compatibility, Unity/master should point to the newest releases sdk, and sdk/master should point to minimum supported version - we want implementers to use the freshest dll, but MRE creators to make MREs that run on every host app). But due to back compat, for now unity/master will point to a version compatible with sdk/master
+
+Minimum supported version=the major.minor version that all host apps support, and have force updated their users to.
+
+Minor version number doesn't *have* to increment in lockstep with altspace's release cycle. We may do more than 1 sdk release between altspace releases.
+
+With this structure, I'd expect there to mostly be 1 or 2 active release versions, but occasionally there would be 3 (or in theory more). The diagram cover 3 active versions for comprehension
+
+(* indicates a detailed workflow below)
+
+## Main workflows:
+### [Orange] Integrate minor unity-only change
+1. Develop feature* targeting unity/red
+
+### [Green] integrate node-only change (or hotfix)
+1. Develop feature* targeting sdk/green, samples/green
+
+### [Red] integrate breaking unity+node change
+1. Develop feature* targeting unity/red, sdk/red, samples/red
+
+### [Blue] When we decide to prepare a new MRE minor version release (criteria: usually ~1 day before Altspace code lock, or sooner if there are major changes)
+1. Merge sdk/green into sdk/red
+2. Merge samples/green into samples/red
+3. Make sure protocol version is incremented in red branches if needed:
+   1. MREUnityRuntimeLib\Constants.cs: ProtocolVersion
+   2. packages\sdk\src\utils\verifyClient.ts: CurrentProtocolVersion
+   3. **TODO: We should change to have both current and min protocol versions. Unity rejects if MRE currentprotocolversion is higher than unity.Current or older than unity.min, and MRE rejects only if Unity.current is older than min (because newer clients are compatible with  older MREs)**
+4. Test pass* for red branches
+5. Create new unity/sdk/sample branches named v0.[new_minor].n, matching the red (NOTE: if there was already a new SDK release created, but it wasn't "shipped", we can stomp that - i.e. if v0.[new_minor-1].n was never used in any host app, we can just reset hard, instead of creating v0.[new_minor].n)
+6. Build Unity DLLs* from unity/v0.[new_minor].n
+7. Publish NPM package* from sdk/v0.[new_minor].n tagged @next
+
+### [Cyan] When we decide to integrate sdk/green into an existing minor release (criteria: whenever we think it's worth the effort)
```

Comment:
It isn't clear to me why we would want to patch an existing minor release. NPM doesn't allow republish of an existing version of a package. If republishing isn't possible, what's the value here? Or is the existing minor release unpublished at this point?

Reply:
"existing minor release" can refer to either to the version that master is pointing to, OR to a version that may become master soon, but is not yet in master. I should clarify that this would be as a Patch onto each of the minor releases 

---
In reply to: [251498412](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/142#discussion_r251498412) [](ancestors = 251498412)

Reply:
clarified 

---
In reply to: [252374993](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/142#discussion_r252374993) [](ancestors = 252374993,251498412)